lib/commons/aria/lookup-table.js

```diff
@@ -2079,6 +2094,132 @@ lookupTable.role = {
 	}
 };
 
+// Source: https://www.w3.org/TR/html-aam-1.0/#element-mapping-table
+// Source: https://www.w3.org/TR/html-aria/
+lookupTable.implicitRole = {
+	a: vNode => {
+		return vNode.hasAttr('href') ? 'link' : null;
+	},
+	area: vNode => {
+		return vNode.hasAttr('href') ? 'link' : null;
+	},
+	article: 'article',
+	aside: 'complementary',
+	body: 'document',
+	button: 'button',
+	datalist: 'listbox',
+	dd: 'definition',
+	dfn: 'term',
+	details: 'group',
+	dialog: 'dialog',
+	dt: 'term',
+	fieldset: 'group',
+	figure: 'figure',
+	footer: vNode => {
+		const sectioningElement = closest(
+			vNode,
+			'article, aside, main, nav, section, [role=article], [role=complementary], [role=main], [role=navigation], [role=region]'
+		);
+
+		return !sectioningElement ? 'contentinfo' : null;
+	},
+	form: vNode => {
+		// accessible-text-virtual uses getRole, so we have to only
+		// use the name calculation steps we know apply to a form
+		// element
+		return arialabelledbyText(vNode) || arialabelText(vNode) || titleText(vNode)
+			? 'form'
+			: null;
+	},
+	h1: 'heading',
+	h2: 'heading',
+	h3: 'heading',
+	h4: 'heading',
+	h5: 'heading',
+	h6: 'heading',
+	header: vNode => {
+		const sectioningElement = closest(
+			vNode,
+			'article, aside, main, nav, section, [role=article], [role=complementary], [role=main], [role=navigation], [role=region]'
+		);
+
+		return !sectioningElement ? 'banner' : null;
+	},
+	hr: 'separator',
+	img: vNode => {
+		return vNode.hasAttr('alt') && !vNode.attr('alt') ? null : 'img';
+	},
+	input: vNode => {
+		// Source: https://www.w3.org/TR/html52/sec-forms.html#suggestions-source-element
+		const listElement = idrefs(vNode.actualNode, 'list').filter(
+			node => !!node
+		)[0];
+		const suggestionsSourceElement =
+			listElement && listElement.nodeName.toLowerCase() === 'datalist';
+
+		switch (vNode.attr('type')) {
+			case 'button':
+			case 'image':
+			case 'reset':
+			case 'submit':
+				return 'button';
+			case 'checkbox':
+				return 'checkbox';
+			case 'email':
+			case 'tel':
+			case 'text':
+			case 'url':
+			case null: // text is the default value
+				return !suggestionsSourceElement ? 'textbox' : 'combobox';
+			case 'number':
+				return 'spinbutton';
+			case 'radio':
+				return 'radio';
+			case 'range':
+				return 'slider';
+			case 'search':
+				return !suggestionsSourceElement ? 'searchbox' : 'combobox';
+		}
```

Comment:
What about if the type isn't any of these? I think that defaults to text, no?

Reply:
`case null:` is that case.